scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/package.scala

```diff
@@ -77,21 +79,69 @@ import scala.collection.JavaConverters._
  *     s(side).getOrElse(x, "unknown")
  *   }
  * }}}
+ *
+ * A `TypedSparkeyReader` can be used to do automatic decoding of JVM types from byte values:
+ * {{{
+ * val main: SCollection[String] = sc.parallelize(Seq("a", "b", "c"))
+ * val side: SideInput[TypedSparkeyReader[MyObject]] = sc
+ *   .typedSparkeySideInput("gs://<bucket>/<path>/<sparkey-prefix>", MyObject.decode)
+ *
+ * val objects: SCollection[MyObject] = main
+ *   .withSideInputs(side)
+ *   .map { (x, s) => s(side).get(x) }
+ *   .toSCollection
+ * }}}
+ *
+ * A `TypedSparkeyReader` can also accept a Caffeine cache to reduce IO and deserialization load:
+ * {{{
+ * val main: SCollection[String] = sc.parallelize(Seq("a", "b", "c"))
+ * val cache: Cache[String, MyObject] = ...
+ * val side: SideInput[TypedSparkeyReader[MyObject]] = sc
+ *   .typedSparkeySideInput("gs://<bucket>/<path>/<sparkey-prefix>", MyObject.decode, cache)
+ *
+ * val objects: SCollection[MyObject] = main
+ *   .withSideInputs(side)
+ *   .map { (x, s) => s(side).get(x) }
+ *   .toSCollection
+ * }}}
  */
 package object sparkey {
 
   /** Enhanced version of [[ScioContext]] with Sparkey methods. */
   implicit class SparkeyScioContext(private val self: ScioContext) extends AnyVal {
 
+    private def viewOf(basePath: String): PCollectionView[SparkeyUri] =
+      self.parallelize(Seq(SparkeyUri(basePath, self.options))).applyInternal(View.asSingleton())
+
     /**
      * Create a SideInput of `SparkeyReader` from a [[SparkeyUri]] base path, to be used with
      * [[com.spotify.scio.values.SCollection.withSideInputs SCollection.withSideInputs]].
      */
-    def sparkeySideInput(basePath: String): SideInput[SparkeyReader] = {
-      val uri = SparkeyUri(basePath, self.options)
-      val view = self.parallelize(Seq(uri)).applyInternal(View.asSingleton())
-      new SparkeySideInput(view)
-    }
+    def sparkeySideInput(basePath: String): SideInput[SparkeyReader] =
+      new SparkeySideInput(viewOf(basePath))
+
+    /**
+     * Create a SideInput of `TypedSparkeyReader` from a [[SparkeyUri]] base path, to be used with
+     * [[com.spotify.scio.values.SCollection.withSideInputs SCollection.withSideInputs]].
+     * The provided decoder function will map from the underlying byte array to a JVM type, and
+     * the optional [[Cache]] object can be used to cache reads in memory after decoding.
+     */
+    def typedSparkeySideInput[T >: Null <: AnyRef](
```

Comment:
This is called in `main` so `cache` has to survive `ser/de`? Also before job execution cache should be empty right? What about make it a cache factory method or something like `f: () => Cache[String, T]`? Or even `SparkeyReader => Cache[String, T]`, so the builder can infer cache settings from sparkey header, e.g. num entries, file size.

For user facing APIs I'd also curry the lambda args so one can do `.typedSSI("...") { b => ... }`. You might have to overload the method for the optional `cache` version.